[GR-47755] Refactor dwarf constants

adinn opened the pull request:
This pull request factors out into interface `DwarfConstants` all the constant values defined in the DWARF spec that are used by the debug info generator code driven via the `DebugInfoProvider` API. This is to allow those same constants to be reused by code that needs to manage DWARF info but is not driven via `DebugInfoProvider`.

The change also cleans up some of the code in the current `DebugInfoProvider` DWARF backend, removing prefix `DW_` from a few constants that are not defined by the DWARF spec and adopting an enum type for the abbreviation codes used to tag DWARF DIEs.

adinn commented:
This is a first step towards factoring out common DWARF management routines as requested in feature request issue #6207.

adinn commented:
@olpaw @fniephaus I pushed a new change to use enums instead of interfaces for the different categories of DWARF constant. So, this is ready for review again (the remaining gate style problem is unrelated to the patch).

fniephaus commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/elf/dwarf/DwarfLineSectionImpl.java:
```diff
@@ -652,15 +593,15 @@ private static SubRange prologueLeafRange(CompiledMethodEntry compiledEntry) {
         int pos = p;
         debugCopyCount++;
         verboseLog(context, "  [0x%08x] Copy %d", pos, debugCopyCount);
         return writeByte(opcode, buffer, pos);
```
Can't we just have a `writeByte(DwarfLineOpcode opcode, ...)` which calls `writeByte(opcode.value(), buffer, pos)`?

adinn replied:
Done.

fniephaus reviewed (approved):
LGTM.

fniephaus commented:
>Once again, the gate failure is a spurious error unrelated to this patch.

It seems you were still working on an older commit. We've fixed this some time ago with https://github.com/oracle/graal/commit/56671ad090fcbc06c8b2121504236c4d8befd713. Rebasing to the latest master branch would've fixed the gate in your PR. But no need to waste free computing resources, I try to get it merged the way it is :)

adinn commented:
Aaargh, so the problem snuck back into my commit. Do you want me to push a fix?

fniephaus commented:
No, it's fine. I've edited your commits to speed things up.

olpaw commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/elf/dwarf/constants/DwarfAccess.java:
```diff
@@ -0,0 +1,46 @@
+    DW_ACCESS_protected((byte) 2),
+    DW_ACCESS_private((byte) 3);
+
+    byte value;
```
`final`

olpaw commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/elf/dwarf/constants/DwarfEncoding.java:
```diff
@@ -0,0 +1,52 @@
+    @SuppressWarnings("unused")
+    DW_ATE_unsigned_char((byte) 0x8);
+
+    byte value;
```
final

olpaw commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/elf/dwarf/constants/DwarfExpressionOpcode.java:
```diff
@@ -0,0 +1,62 @@
+    DW_OP_implicit_value((byte) 0x9e),
+    DW_OP_stack_value((byte) 0x9f);
+
+    byte value;
```
final

olpaw commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/elf/dwarf/constants/DwarfFlag.java:
```diff
@@ -0,0 +1,46 @@
+    DW_FLAG_false((byte) 0),
+    DW_FLAG_true((byte) 1);
+
+    byte value;
```
final

olpaw commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/elf/dwarf/constants/DwarfForm.java:
```diff
@@ -0,0 +1,66 @@
+    DW_FORM_strp(0xe),
+    DW_FORM_expr_loc(0x18);
+
+    int value;
```
final

olpaw commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/elf/dwarf/constants/DwarfFrameValue.java:
```diff
@@ -0,0 +1,68 @@
+    DW_CFA_def_cfa_register((byte) 0xd),
+    DW_CFA_def_cfa_offset((byte) 0xe);
+
+    byte value;
```
final

olpaw commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/elf/dwarf/constants/DwarfHasChildren.java:
```diff
@@ -0,0 +1,45 @@
+    DW_CHILDREN_no((byte) 0),
+    DW_CHILDREN_yes((byte) 1);
+
+    byte value;
```
final

olpaw commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/elf/dwarf/constants/DwarfInline.java:
```diff
@@ -0,0 +1,50 @@
+    @SuppressWarnings("unused")
+    DW_INL_declared_inlined((byte) 3);
+
+    byte value;
```
final

olpaw commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/elf/dwarf/constants/DwarfLanguage.java:
```diff
@@ -0,0 +1,45 @@
+public enum DwarfLanguage {
+    DW_LANG_Java((byte) 0xb);
+
+    byte value;
```
final

olpaw commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/elf/dwarf/constants/DwarfLineOpcode.java:
```diff
@@ -0,0 +1,113 @@
+     */
+    DW_LNE_define_file((byte) 3);
+
+    byte value;
```
final

olpaw commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/elf/dwarf/constants/DwarfSectionName.java:
```diff
@@ -0,0 +1,53 @@
+    DW_ARANGES_SECTION(".debug_aranges"),
+    DW_RANGES_SECTION(".debug_ranges");
+
+    String value;
```
final

olpaw commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/elf/dwarf/constants/DwarfTag.java:
```diff
@@ -0,0 +1,62 @@
+    DW_TAG_unspecified_type(0x3b),
+    DW_TAG_inlined_subroutine(0x1d);
+
+    int value;
```
final

olpaw commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/elf/dwarf/constants/DwarfVersion.java:
```diff
@@ -0,0 +1,51 @@
+    DW_VERSION_2((short) 2),
+    DW_VERSION_4((short) 4);
+
+    short value;
```
final

olpaw commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/elf/dwarf/DwarfSectionImpl.java:
```diff
@@ -55,7 +61,13 @@
 import java.util.stream.Stream;
 
-import static com.oracle.objectfile.elf.dwarf.DwarfDebugInfo.DW_OP_stack_value;
+import static com.oracle.objectfile.elf.dwarf.constants.DwarfSectionName.DW_ABBREV_SECTION;
```
Instead of importing individual enum values use a single
```
import com.oracle.objectfile.elf.dwarf.constants.DwarfSectionName
```

adinn replied:
If you look further up the file you will see that I already have a single

    import com.oracle.objectfile.elf.dwarf.constants.DwarfSectionName

The reason I am using the static imports is that this avoids the need to have every mention of a constant qualified by the enum it belongs to e.g. at line 285 the current code is

        markRelocationSite(pos, ObjectFile.RelocationKind.DIRECT_8, TEXT_SECTION.value(), l);

If I remove the static import I need to change this to

        markRelocationSite(pos, ObjectFile.RelocationKind.DIRECT_8, DwarfSectionName.TEXT_SECTION.value(), l);

So, your suggestion that I remove the static imports implies that I would also need to change every such mention of an enum value so that it is qualified with the name of the enum class it belongs to.

Is that really what you want?

olpaw replied:
> So, your suggestion that I remove the static imports implies that I would also need to change every such mention of an enum value so that it is qualified with the name of the enum class it belongs to.

Yes, that's completely fine. BTW, intellij can do that automatically for you.

![image](https://github.com/oracle/graal/assets/16901644/ec5285b7-4f58-4c6a-b077-5bafb3f665ac)

adinn replied:
Yes, I understand that this can be done automatically. I wanted to be sure this is what you want. Personally, I would *much rather* avoid having to include the enum class qualifiers as:

1. it adds needless clutter to the code (readability--)
2. that clutter clouds the otherwise obvious and immediate implication that that these symbols are precisely those defined by the DWARF spec (meaningfulness--)
3. qualification of the names will not automatically result in a compile failure should someone by any chance attempt to reuse the DWARF spec names (reliability--).

However, if you are sure you really prefer it this way then I will change it.

olpaw replied:
It's a tradeoff. Having a huge block of `static imports` is also not pretty. And every time you are using a new enum value another one is added. There is a tipping point where I also opt for static imports over qualified access. E.g. if there are many many places where I would need that one single enum value. But this is not really the case here so I do in fact prefer a import + qualified access wherever a value is used.

olpaw reviewed (changes requested):
Please replace all the individual `import static` of enum values with single regular imports of the respective enum classes.

adinn commented:
I have pushed a patch that 1) makes all the enum values private and 2) removes all static enum value imports, instead resorting to enum type qualified mentions at the point of use. I think that addresses all the review comments.

adinn commented:
Rebased over latest master

fniephaus commented:
>Rebased over latest master

Thanks! `HybridLayoutSupport.java` is still modified in https://github.com/oracle/graal/pull/7081/commits/2ceefd58b3eb53b9205a0e7b8f97601cc9d2b9a5#diff-6f4e2943f89b429e13da3d3264be1c672561950b119cfabd34467382634d1929 but I'll take care of it...the PR is now in the merge queue and (unfortunately) duplicated at #7158.

fniephaus commented:
Merged via https://github.com/oracle/graal/pull/7158#event-10042994039.
